controllers/machine_controller_test.go

```diff
@@ -29,16 +28,236 @@ import (
 	"k8s.io/client-go/kubernetes/scheme"
 	"k8s.io/utils/pointer"
 	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/client/fake"
 	"sigs.k8s.io/controller-runtime/pkg/log"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
-var _ = Describe("Machine Reconciler", func() {
-	It("Should create a Machine", func() {
-		// TODO
-	})
-})
+func TestMachineFinalizer(t *testing.T) {
+	clusterCorrectMeta := &clusterv1.Cluster{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Cluster",
+			APIVersion: clusterv1.GroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "default",
+			Name:      "valid-cluster",
+		},
+	}
+
+	machineValidCluster := &clusterv1.Machine{
+		TypeMeta: metav1.TypeMeta{Kind: "Machine"},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "machine1",
+			Namespace: "default",
+			Labels: map[string]string{
+				clusterv1.MachineClusterLabelName: "valid-cluster",
+			},
+		},
+	}
+
+	machineWithFinalizer := &clusterv1.Machine{
+		TypeMeta: metav1.TypeMeta{Kind: "Machine"},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "machine2",
+			Namespace: "default",
+			Labels: map[string]string{
+				clusterv1.MachineClusterLabelName: "valid-cluster",
+			},
+			Finalizers: []string{"some-other-finalizer"},
+		},
+	}
+
+	testCases := []struct {
+		name               string
+		request            reconcile.Request
+		m                  *clusterv1.Machine
+		expectedFinalizers []string
+	}{
+		{
+			name: "should add a machine finalizer to the machine if it doesn't have one",
+			request: reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      machineValidCluster.Name,
+					Namespace: machineValidCluster.Namespace,
+				},
+			},
+			m:                  machineValidCluster,
+			expectedFinalizers: []string{clusterv1.MachineFinalizer},
+		},
+		{
+			name: "should append the machine finalizer to the machine if it already has a finalizer",
+			request: reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      machineWithFinalizer.Name,
+					Namespace: machineWithFinalizer.Namespace,
+				},
+			},
+			m:                  machineWithFinalizer,
+			expectedFinalizers: []string{"some-other-finalizer", clusterv1.MachineFinalizer},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			RegisterTestingT(t)
+			clusterv1.AddToScheme(scheme.Scheme)
+			mr := &MachineReconciler{
+				Client: fake.NewFakeClientWithScheme(
+					scheme.Scheme,
```

Comment:
The reason I used this method was because `fake.NewFakeClient` is [marked as deprecated](https://github.com/kubernetes-sigs/controller-runtime/blob/aac8bc7a3cbc56b99cc250d1506a0a39765b573d/pkg/client/fake/client.go#L44-L50).

Reply:
@vincepri Also if you are referring to the scheme being initialized in the [`suite_test.go` in the `BeforeSuite`](https://github.com/kubernetes-sigs/cluster-api/blob/7f1890e4ebd9387491493b8018470b1155bba018/controllers/suite_test.go#L89), I assumed that this would be run as part of the ginkgo/integration tests and not for the unit tests when running `go test ./...`

Reply:
Good to know, thanks!